sql: TestSqlActivityJobRunsAfterStatsFlush failed

sql.TestSqlActivityJobRunsAfterStatsFlush [failed](https://tanzanite.cluster.engflow.com/invocations/default/2e180b8d-c74c-448d-8b1a-149448c4ebcb?testReportRun=1&testReportShard=14&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [79e6cec76716b179d6982ac9b969c117f6644ea9](https://github.com/cockroachdb/cockroach/commits/79e6cec76716b179d6982ac9b969c117f6644ea9):

Fatal error:

```
panic: test timed out after 40m0s
	running tests:
		TestSqlActivityJobRunsAfterStatsFlush (37m29s)

```
Stack: 

```
goroutine 28308230 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2373 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:215 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* 
* goroutine 11674421 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs.func1()
* 	pkg/util/admission/work_queue.go:501 +0x14d
* created by github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs in goroutine 11672324
* 	pkg/util/admission/work_queue.go:472 +0x4f
* 
* goroutine 11675627 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble/internal/genericcache.(*shard[...]).releaseLoop(0xc01384a748)
* 	external/com_github_cockroachdb_pebble/internal/genericcache/shard.go:61 +0x6f
* created by github.com/cockroachdb/pebble/internal/genericcache.(*shard[...]).Init in goroutine 11672324
* 	external/com_github_cockroachdb_pebble/internal/genericcache/shard.go:54 +0x1a7
* 
* goroutine 11685405 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue.func2()
* 	pkg/util/admission/work_queue.go:424 +0x85
* created by github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue in goroutine 11672324
* 	pkg/util/admission/work_queue.go:421 +0x3be
* 
* goroutine 11677322 [select, 2 minutes]:
* github.com/cockroachdb/cockroach/pkg/util/hlc.(*Clock).StartMonitoringForwardClockJumps.func1()
* 	pkg/util/hlc/hlc.go:298 +0x13d
* created by github.com/cockroachdb/cockroach/pkg/util/hlc.(*Clock).StartMonitoringForwardClockJumps in goroutine 11672324
* 	pkg/util/hlc/hlc.go:290 +0x106
* 
* goroutine 11674487 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue.func2()
* 	pkg/util/admission/work_queue.go:424 +0x85
* created by github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue in goroutine 11672324
* 	pkg/util/admission/work_queue.go:421 +0x3be
* 
* goroutine 11720451 [semacquire, 2 minutes]:
* sync.runtime_Semacquire(0x10?)
* 	GOROOT/src/runtime/sema.go:71 +0x25
* sync.(*WaitGroup).Wait(0xc01d7a9b20?)
* 	GOROOT/src/sync/waitgroup.go:118 +0x48
* github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).flush(0xc01386ad80, {0x85ebda8, 0xc01677f500}, 0xc0187d55f0, 0xc0143b2fc0, {0xc013abc600, 0xb, 0xc}, {0xc017596a20, 0x7, ...})
* 	pkg/sql/sqlstats/persistedsqlstats/flush.go:216 +0x432
* github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).MaybeFlushWithDrainer(0xc01386ad80, {0x85ebda8, 0xc01677f500}, 0xc0187d55f0, {0x85a8178, 0xc014d0c5a0})
* 	pkg/sql/sqlstats/persistedsqlstats/flush.go:115 +0x88c
* github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).MaybeFlush(...)
* 	pkg/sql/sqlstats/persistedsqlstats/flush.go:52
* github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).startSQLStatsFlushLoop.func1({0x85ebda8, 0xc01677f500})
* 	pkg/sql/sqlstats/persistedsqlstats/provider.go:177 +0x2d5
* github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x7613a8a, 0x10})
* 	pkg/util/stop/stopper.go:500 +0x1e2
* created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 11672324
* 	pkg/util/stop/stopper.go:489 +0x466
* 
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=14</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSqlActivityJobRunsAfterStatsFlush.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-48153

--- a/pkg/sql/sqlstats/persistedsqlstats/flush.go
+++ b/pkg/sql/sqlstats/persistedsqlstats/flush.go
@@ -200,6 +200,8 @@ func (s *PersistedSQLStats) flush(
 	})
 	if err != nil {
 		log.Warningf(ctx, "failed to execute sql-stmt-stats-flush task, %s", err.Error())
+		wg.Done()
+		return
 	}
 
 	err = stopper.RunAsyncTask(ctx, "sql-txn-stats-flush", func(ctx context.Context) {
@@ -211,6 +213,8 @@ func (s *PersistedSQLStats) flush(
 	})
 	if err != nil {
 		log.Warningf(ctx, "failed to execute sql-txn-stats-flush task, %s", err.Error())
+		wg.Done()
+		return
 	}
 
 	wg.Wait()

